core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+    pub(super) node_count: u32,
+}
+
+impl Flush for BTreeMapHeader {
+    #[inline]
+    fn flush(&mut self) {
+        self.next_vacant.flush();
+        self.root.flush();
+        self.len.flush();
+        self.node_count.flush();
+    }
+}
+
+/// A node in the tree.
+///
+/// Each node contains `CAPACITY` keys and values and an edges array over
+/// which children nodes can be linked. Each child has a link to its parent.
+///
+/// Each node is stored as one storage entity. This reduces storage access,
+/// since with each fetch the entire content of a node (all its elements, etc.)
+/// are fetched.
+#[derive(PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct Node<K, V> {
+    /// This node's parent node.
+    pub parent: Option<u32>,
+
+    /// This node's index into the parent node's `edges` array.
+    /// If, for example, `parent_idx = Some(2)` this refers to the
+    /// second position in the `edges` array of its parent node.
+    pub parent_idx: Option<u32>,
+
+    /// The array storing the keys for a node.
+    pub keys: [Option<K>; CAPACITY],
+
+    /// The array storing the values for a node.
+    pub vals: [Option<V>; CAPACITY],
+
+    /// The pointers to the children of this node.
+    pub edges: [Option<u32>; 2 * B],
+
+    /// Number of elements stored in this node.
+    pub len: u32,
+}
+
+impl<K, V> Flush for Node<K, V>
+where
+    K: Encode + Flush,
+    V: Encode + Flush,
+{
+    #[inline]
+    fn flush(&mut self) {
+        self.parent.flush();
+        self.parent_idx.flush();
+        self.keys.flush();
+        self.vals.flush();
+        self.edges.flush();
+        self.len.flush();
+    }
+}
+
+impl<K, V> Node<K, V> {
+    /// Creates a new `Node`. The node is empty and all fields are
+    /// filled with `None`.
+    pub fn new() -> Self {
+        Node {
+            parent: None,
+            parent_idx: None,
+            keys: Default::default(),
+            vals: Default::default(),
+            edges: [None; 2 * B],
+            len: 0,
+        }
+    }
+
+    /// Returns the number of elements stored in the tree.
+    pub fn len(&self) -> usize {
+        self.len as usize
+    }
+
+    /// Returns the number of edges this node has.
+    pub fn edges(&self) -> usize {
+        self.edges.iter().filter(|o| o.is_some()).count()
+    }
+}
+
+/// Points to a node in the tree.
+pub struct NodeHandle {
+    node: u32,
+}
+
+impl NodeHandle {
+    pub fn new(node: u32) -> Self {
+        Self { node }
+    }
+
+    pub fn node(&self) -> u32 {
+        self.node
+    }
+}
+
+impl From<&KVHandle> for NodeHandle {
+    fn from(handle: &KVHandle) -> NodeHandle {
+        NodeHandle::new(handle.node)
+    }
+}
+
+impl From<KVHandle> for NodeHandle {
+    fn from(handle: KVHandle) -> NodeHandle {
+        NodeHandle::new(handle.node)
+    }
+}
+
+impl From<u32> for NodeHandle {
+    fn from(index: u32) -> NodeHandle {
+        NodeHandle::new(index)
+    }
+}
```

Comment:
Sure you want this? Because a trait implementation is always public. But for improved type safety by using index spaces guided by index types it makes sense to keep implementations like these as private as possible.